Add tree favorites

maurizi opened the pull request:
Builds on top of #1948

Adds a UI for favoriting trees, and converts the existing `TreeFavorite` model into a map feature based `Favorite` model.

maurizi commented:
@steventlamb I addressed some of your comments.
If you still have concerns about the OTM1 migrator, I would appreciate some help.

steventlamb commented:
Pulled down and tested, but I can't run the migrations
```
[127.0.0.1] out:  > treemap:0092_add_map_feature_to_tree_favorite
[127.0.0.1] out: Error in migration: treemap:0092_add_map_feature_to_tree_favorite
[127.0.0.1] out: Traceback (most recent call last):
[127.0.0.1] out:   File "manage.py", line 10, in <module>
[127.0.0.1] out:     execute_from_command_line(sys.argv)
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 399, in 
execute_from_command_line                                                                                                      
[127.0.0.1] out:     utility.execute()
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 392, in 
execute                                                                                                                        
[127.0.0.1] out:     self.fetch_command(subcommand).run_from_argv(self.argv)
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/core/management/base.py", line 242, in run_
from_argv                                                                                                                      
[127.0.0.1] out:     self.execute(*args, **options.__dict__)
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/core/management/base.py", line 285, in exec
ute                                                                                                                            
[127.0.0.1] out:     output = self.handle(*args, **options)
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/management/commands/migrate.py", line 108, i
n handle                                                                                                                       
[127.0.0.1] out:     ignore_ghosts = ignore_ghosts,
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/__init__.py", line 213, in migrate
_app                                                                                                                           
[127.0.0.1] out:     success = migrator.migrate_many(target, workplan, database)
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 235, in migrat
e_many                                                                                                                         
[127.0.0.1] out:     result = migrator.__class__.migrate_many(migrator, target, migrations, database)
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 310, in migrat
e_many                                                                                                                         
[127.0.0.1] out:     result = self.migrate(migration, database)
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 133, in migrat
e                                                                                                                              
[127.0.0.1] out:     result = self.run(migration)
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 107, in run
[127.0.0.1] out:     return self.run_migration(migration)
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 81, in run_mig
ration                                                                                                                         
[127.0.0.1] out:     migration_function()
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 57, in <lambda
>                                                                                                                              
[127.0.0.1] out:     return (lambda: direction(orm))
[127.0.0.1] out:   File "/usr/local/otm/app/opentreemap/treemap/migrations/0092_add_map_feature_to_tree_favorite.py", line 10, i
n forwards                                                                                                                     
[127.0.0.1] out:     for favorite in orm.TreeFavorite.objects.all():
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 96, in __iter__
[127.0.0.1] out:     self._fetch_all()
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 854, in _fetch_al
l                                                                                                                              
[127.0.0.1] out:     self._result_cache = list(self.iterator())
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 220, in iterator
[127.0.0.1] out:     for row in compiler.results_iter():
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 710, in re
sults_iter                                                                                                                     
[127.0.0.1] out:     for rows in self.execute_sql(MULTI):
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 781, in ex
ecute_sql                                                                                                                      
[127.0.0.1] out:     cursor.execute(sql, params)
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/db/backends/util.py", line 69, in execute
[127.0.0.1] out:     return super(CursorDebugWrapper, self).execute(sql, params)
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/db/backends/util.py", line 53, in execute
[127.0.0.1] out:     return self.cursor.execute(sql, params)
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/db/utils.py", line 99, in __exit__
[127.0.0.1] out:     six.reraise(dj_exc_type, dj_exc_value, traceback)
[127.0.0.1] out:   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/db/backends/util.py", line 53, in execute
[127.0.0.1] out:     return self.cursor.execute(sql, params)
[127.0.0.1] out: django.db.utils.ProgrammingError: column treemap_treefavorite.created does not exist
[127.0.0.1] out: LINE 1: SELECT "treemap_treefavorite"."created", "treemap_treefavori...
[127.0.0.1] out:                ^
[127.0.0.1] out: 
[127.0.0.1] out: 


Fatal error: sudo() received nonzero return code 1 while executing!

Requested: /usr/local/otm/env/bin/python manage.py migrate --noinput
Executed: sudo -S -p 'sudo password:'  /bin/bash -l -c "cd /usr/local/otm/app/opentreemap && /usr/local/otm/env/bin/python manag
e.py migrate --noinput"                                                                                                        

Aborting.
Disconnecting from 127.0.0.1:2222... done.
```

jwalgran commented on opentreemap/treemap/templates/treemap/map_feature_detail.html:
```diff
@@ -1,4 +1,4 @@
-{% extends "instance_base.html" %}
+<div>% extends "instance_base.html" %</div>
```
This template tag needs to be fixed

maurizi replied:
Oops, thanks

jwalgran commented:
If I load the detail page, then stop my dev server, clicking the star does nothing. I would like to add some error toast, or add an issue to add some error toast in the future.

jwalgran commented:
+1 I can toggle the favorite and it persists.

jwalgran commented:
I added a favorite and was unable to rollback the migrations. I removed the favorite, and the rollback worked. Not sure how or if we want to handle this.

```
$ python manage.py migrate treemap 0090
 - Soft matched migration 0090 to 0090_auto__del_field_species_gender.
Running migrations for treemap:
 - Migrating backwards to just after 0090_auto__del_field_species_gender.
 < treemap:0094_auto__rename_treefavorite_favorite
 < treemap:0093_auto__del_field_treefavorite_tree__chg_field_treefavorite_map_feature_
 < treemap:0092_add_map_feature_to_tree_favorite
Error in migration: treemap:0092_add_map_feature_to_tree_favorite
Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 399, in execute_from_command_line
    utility.execute()
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 392, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/core/management/base.py", line 242, in run_from_argv
    self.execute(*args, **options.__dict__)
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/core/management/base.py", line 285, in execute
    output = self.handle(*args, **options)
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/management/commands/migrate.py", line 108, in handle
    ignore_ghosts = ignore_ghosts,
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/__init__.py", line 213, in migrate_app
    success = migrator.migrate_many(target, workplan, database)
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 235, in migrate_many
    result = migrator.__class__.migrate_many(migrator, target, migrations, database)
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 356, in migrate_many
    self.migrate(migration, database)
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 133, in migrate
    result = self.run(migration)
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 107, in run
    return self.run_migration(migration)
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 81, in run_migration
    migration_function()
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 57, in <lambda>
    return (lambda: direction(orm))
  File "/usr/local/otm/app/opentreemap/treemap/migrations/0092_add_map_feature_to_tree_favorite.py", line 18, in backwards
    favorite.tree = orm.Tree.objects.get(plot_id=feature.id)
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/db/models/manager.py", line 151, in get
    return self.get_queryset().get(*args, **kwargs)
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 307, in get
    self.model._meta.object_name)
treemap.models.DoesNotExist: Tree matching query does not exist.

maurizi commented:
@jwalgran To be honest I'm inclined to not bother with backwards migrations.
On Feb 11, 2015 6:00 PM, "Justin Walgran" <notifications@github.com> wrote:

> I added a favorite and was unable to rollback the migrations. I removed
> the favorite, and the rollback worked. Not sure how or if we want to handle
> this.
>
> $ python manage.py migrate treemap 0090
>  - Soft matched migration 0090 to 0090_auto__del_field_species_gender.
> Running migrations for treemap:
>  - Migrating backwards to just after 0090_auto__del_field_species_gender.
>  < treemap:0094_auto__rename_treefavorite_favorite
>  < treemap:0093_auto__del_field_treefavorite_tree__chg_field_treefavorite_map_feature_
>  < treemap:0092_add_map_feature_to_tree_favorite
> Error in migration: treemap:0092_add_map_feature_to_tree_favorite
> Traceback (most recent call last):
>   File "manage.py", line 10, in <module>
>     execute_from_command_line(sys.argv)
>   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 399, in execute_from_command_line
>     utility.execute()
>   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 392, in execute
>     self.fetch_command(subcommand).run_from_argv(self.argv)
>   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/core/management/base.py", line 242, in run_from_argv
>     self.execute(*args, **options.__dict__)
>   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/core/management/base.py", line 285, in execute
>     output = self.handle(*args, **options)
>   File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/management/commands/migrate.py", line 108, in handle
>     ignore_ghosts = ignore_ghosts,
>   File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/__init__.py", line 213, in migrate_app
>     success = migrator.migrate_many(target, workplan, database)
>   File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 235, in migrate_many
>     result = migrator.__class__.migrate_many(migrator, target, migrations, database)
>   File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 356, in migrate_many
>     self.migrate(migration, database)
>   File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 133, in migrate
>     result = self.run(migration)
>   File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 107, in run
>     return self.run_migration(migration)
>   File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 81, in run_migration
>     migration_function()
>   File "/usr/local/otm/env/local/lib/python2.7/site-packages/south/migration/migrators.py", line 57, in <lambda>
>     return (lambda: direction(orm))
>   File "/usr/local/otm/app/opentreemap/treemap/migrations/0092_add_map_feature_to_tree_favorite.py", line 18, in backwards
>     favorite.tree = orm.Tree.objects.get(plot_id=feature.id)
>   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/db/models/manager.py", line 151, in get
>     return self.get_queryset().get(*args, **kwargs)
>   File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 307, in get
>     self.model._meta.object_name)
> treemap.models.DoesNotExist: Tree matching query does not exist.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OpenTreeMap/OTM2/pull/1949#issuecomment-73983974>.
>
